modules/javafx.graphics/src/main/java/javafx/scene/Node.java

```diff
@@ -8171,22 +8174,38 @@ final void notifyFocusListeners() {
     }
 
     /**
-     * Called when the current node was removed from the scene graph.
-     * If the current node has the focusWithin bit, we also need to clear the focusWithin bits of this
-     * node's parents. Note that a scene graph can have more than a single focused node, for example when
-     * a PopupWindow is used to present a branch of the scene graph. Since we need to preserve multi-level
+     * Called when the current node was removed from or added to the scene graph.
+     * If the current node has the focusWithin bit, we also need to clear and set the focusWithin bits of this
+     * node's old and new parents. Note that a scene graph can have more than a single focused node, for example
+     * when a PopupWindow is used to present a branch of the scene graph. Since we need to preserve multi-level
      * focus, we need to adjust the focus-within count on all parents of the node.
      */
-    private void clearParentsFocusWithin(Node oldParent) {
-        if (oldParent != null && focusWithin.get()) {
-            Node node = oldParent;
-            while (node != null) {
-                node.focusWithin.adjust(-focusWithin.count);
-                node = node.getParent();
-            }
+    private void updateParentsFocusWithin(Node oldParent, Node newParent) {
+        if (!focusWithin.get()) {
+            return;
+        }
+
+        Node node = oldParent;
+        while (node != null) {
+            node.focusWithin.adjust(-focusWithin.count);
```

Comment:
suppose we are removing a top level Pane with a bunch of Nodes in it that have non-zero focusWithin.count values.
the parent of said Pane gets a zero count (which is correct), whereas the children of said Pane have their counts unchanged (I assume).

what happens when the same Pane gets added back to the scene graph?  wouldn't we enter an inconsistent state?

Reply:
> suppose we are removing a Pane with a bunch of Nodes in it that have non-zero focusWithin.count values. the parent of said Pane gets a zero count (which is correct), whereas the children of said Pane have their counts unchanged (I assume).

Correct.

> what happens when the same Pane gets added back to the scene graph? wouldn't we enter an inconsistent state?

That's what this patch fixes. If we later add the pane that contains a number of focused nodes back to the scene graph, the `focusWithin` count of its new parents (which was 0 before) is now incremented to match the number of contained focused nodes. The code starts in L8194.

Reply:
thank you for explanations!

one more question: suppose, prior to removal of the Pane it contained a focused (Scene.focusOwner) Node.
1. once the Pane is removed, what happens to Scene.focusOwner?
2. once the Pane is added back, would we have an inconsistent state in respect to Scene.focusOwner (if some other node requested focus in the mean time, wouldn't we have two focused nodes?)

Reply:
> thank you for explanations!
> 
> one more question: suppose, prior to removal of the Pane it contained a focused (Scene.focusOwner) Node.
> 
> 1. once the Pane is removed, what happens to Scene.focusOwner?

There's quite a bit of logic happening. `Scene.focusCleanup()` is a good starting point. Basically, we need to find a new focus owner if the old focus owner is removed from the scene graph.

> 2. once the Pane is added back, would we have an inconsistent state in respect to Scene.focusOwner (if some other node requested focus in the mean time, wouldn't we have two focused nodes?)

It's not inconsistent, but it wouldn't be the same state as before. The `focused` state would remain, but `Scene.focusOwner` would not be restored to the previous node. That part of the focus story hasn't changed with the introduction of `focus-visible` and `focus-within`.

Reply:
So, in the scenario # 2, we would arrive at situation when the previously focused Node is still decorated as having the focus, but won't be getting the events since Scene.focusOwner is different?  I suppose it's the way the whole thing is designed to work.
Thank you for the explanations!